va/validation-authority.go

```diff
@@ -559,27 +507,31 @@ func (va *ValidationAuthorityImpl) validate(authz core.Authorization, challengeI
 	va.RA.OnValidationUpdate(authz)
 }
 
-func (va *ValidationAuthorityImpl) validateChallengeAndCAA(identifier core.AcmeIdentifier, challenge core.Challenge, regID int64) (core.Challenge, error) {
+func (va *ValidationAuthorityImpl) validateChallengeAndCAA(identifier core.AcmeIdentifier, challenge core.Challenge, regID int64) ([]core.ValidationRecord, *probs.ProblemDetails) {
 	ch := make(chan *probs.ProblemDetails, 1)
 	go func() {
 		ch <- va.checkCAA(identifier, regID)
 	}()
 
-	result, err := va.validateChallenge(identifier, challenge)
+	validationRecords, err := va.validateChallenge(identifier, challenge)
 	if err != nil {
-		return result, err
+		return validationRecords, err
 	}
 
-	problemDetails := <-ch
-	if problemDetails != nil {
-		result.Error = problemDetails
-		result.Status = core.StatusInvalid
-		return result, problemDetails
+	caaProblem := <-ch
+	if caaProblem != nil {
+		return validationRecords, caaProblem
 	}
-	return result, nil
+	return validationRecords, nil
 }
 
-func (va *ValidationAuthorityImpl) validateChallenge(identifier core.AcmeIdentifier, challenge core.Challenge) (core.Challenge, error) {
+func (va *ValidationAuthorityImpl) validateChallenge(identifier core.AcmeIdentifier, challenge core.Challenge) ([]core.ValidationRecord, *probs.ProblemDetails) {
+	if !challenge.IsSane(true) {
```

Comment:
My goal was to consolidate the places that can return an error / set a challenge to invalid. Keeping this in `validate` would have meant one extra place where we set status = invalid.